### PR TITLE
Automatic generation for entrypoint code.

### PR DESCRIPTION
**Pull request by guumaster**

### CODE REVIEW

File: runtime-entrypoint/Dockerfile [hunk 1]
```diff
@@ -1,9 +1,27 @@
-FROM python:3.7.5
+FROM ubuntu:18.04
```
Comment (angelsanchez) [inline runtime-entrypoint/Dockerfile:1]: why should be we use ubuntu instead of python Docker image?

File: runtime-entrypoint/src/kre_grpc.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        except Exception as err:
+            err_msg = f'Exception on gRPC call : {err})'
+            self.logger.error(err_msg)
+            traceback.print_exc()
+
+            response = self.make_response_object(subject, KreNatsMessage(error=err_msg))
+            await stream.send_message(response)
+
+
+def message_to_dict(message):
```
Comment (angelsanchez) [inline runtime-entrypoint/src/kre_grpc.py:74]: is this function doing something similar to the installed dependency `protobuf3-to-dict = "0.1.5"`?

File: runtime-entrypoint/Dockerfile (hunk 1 above)
Comment (guumaster) [reply runtime-entrypoint/Dockerfile:15]: we decided to keep ubuntu until we can figure out why pipenv doesn't install dependencies correctly.

File: runtime-entrypoint/src/kre_grpc.py (hunk 2 above)
Comment (guumaster) [reply runtime-entrypoint/src/kre_grpc.py]: good catch. updated to library code instead of custom function.

File: krt-template/run_test.sh [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+DATA_FILE=${3:-"./test/greeter.json"}
+
+ENTRYPOINT_SERVICE=entrypoint.Entrypoint/Greet
+
+
+DEFAULT_MESSAGES=1
+TOTAL_MESSAGES=${3:-$DEFAULT_MESSAGES}
+SEND_CONCURRENT=1
+
+NAMESPACE="kre-$RUNTIME" # NOTE: Runtime on KRE must be called 'broadcom'
```
Comment (angelsanchez) [inline krt-template/run_test.sh:18]:
```suggestion
NAMESPACE="kre-$RUNTIME"
```

File: krt-template/run_test.sh [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+
+run_test() {
+  check_not_empty "VERSION" "missing version name"
+
+  echo_info_header "sending ${TOTAL_MESSAGES} images"
```
Comment (angelsanchez) [inline krt-template/run_test.sh:24]:
```suggestion
  echo_info_header "sending ${TOTAL_MESSAGES} messages"
```

File: krt-template/run_test.sh [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+  ## CALL ENTRYPOINT FOR EACH IMAGE
```
Comment (angelsanchez) [inline krt-template/run_test.sh:26]:
```suggestion
  ## CALL ENTRYPOINT FOR EACH MESSAGE
```

File: runtime-runners/kre-py/src/main.py [hunk 6]
```diff
@@ -217,49 +61,56 @@ def load_handler_module(self):
...
+                output_subject = request_msg.reply if is_last_node else self.config.nats_output
+                response_msg = KreNatsMessage(reply=request_msg.reply, data=handler_result)
 
-                await self.nc.publish(output_subject, output_result.to_json())
-                self.logger.info(f"published response to '{output_subject}' subject with final reply '{result.reply}'")
+                await self.nc.publish(output_subject, response_msg.marshal())
+                self.logger.info(f"published response to '{output_subject}' with final reply '{request_msg.reply}'")
 
-                await self.nc.flush(timeout=NATS_FLUSH_TIMEOUT)
+                await self.nc.flush(timeout=self.config.nats_flush_timeout)
```
Comment (angelsanchez) [inline runtime-runners/kre-py/src/main.py:111]: I got the following error running the greeting krt:
```
2020-07-03T07:10:31+0000 INFO published response to '_INBOX.bGyEoavEO3snXVNnLh9YuV.bGyEoavEO3snEZNnLh9YuV' with final reply '_INBOX.bGyEoavEO3snXVNnLh9YuV.bGyEoavEO3snEZNnLh9YuV'                                
 Traceback (most recent call last):                                                                                                                                                                                
   File "/kre/src/main.py", line 108, in message_cb                                                                                                                                                                
     await self.nc.flush(timeout=self.config.nats_flush_timeout)                                                                                                                                                   
 AttributeError: 'Config' object has no attribute 'nats_flush_timeout'                                                                                                                                             
 2020-07-03T07:10:31+0000 ERROR error executing handler:'Config' object has no attribute 'nats_flush_timeout'
```

File: runtime-entrypoint/entrypoint-gen/entrypoint.py.tmpl [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    {{ range .Methods }}
+    async def {{ .Name }}(self, stream: Stream[{{ .RequestType }}, {{ .ReturnsType }}]) -> None:
+        return await self.process_message(stream, "{{ .Name }}")
+    {{ end }}
+
+    def make_response_object(self, subject, kre_nats_msg):
+        {{- range .Methods }}
+        if subject == '{{ .Name }}':
+            self.logger.info(f"call to {{ .Name }}({{ .RequestType }})"
+                             f" responses with '{{ .ReturnsType }}'. Data: {kre_nats_msg.data}")
```
Comment (angelsanchez) [inline runtime-entrypoint/entrypoint-gen/entrypoint.py.tmpl:23]: Should we log the complete `kre_nats_msg.data`?
Consider cases in which the data is a big amount of information.

File: runtime-entrypoint/entrypoint.sh [hunk 8]
```diff
@@ -1,3 +1,27 @@
...
+#!/bin/sh
 
-sh -c "python $KRT_ENTRYPOINT 2>&1 | tee -a /var/log/app/app.log"
+set -eu
+
+PROTO_FILE="/krt-files/public_input.proto"
+ENTRYPOINT_FILE="/app/src/entrypoint/entrypoint.py"
+CONTROL_FILE="/app/src/entrypoint/.generated"
+
+if [ ! -f "$CONTROL_FILE" ];then
```
Comment (angelsanchez) [inline runtime-entrypoint/entrypoint.sh:9]: Are you sure this is working?
After starting/stopping versions I always see the following logs:
```generating '/app/src/entrypoint/entrypoint.py' file.                                                                                                                                                              
 Done                                                                                                                                                                                                              
 generating entrypoint protofub files.                                                                                                                                                                             
 2020-07-03T07:21:48+0000 INFO Connecting to NATS kre-nats:4222...```

File: runtime-entrypoint/entrypoint-gen/entrypoint.py.tmpl (hunk 7 above)
Comment (guumaster) [reply runtime-entrypoint/entrypoint-gen/entrypoint.py.tmpl:23]: changed to log only length and type.

File: runtime-runners/kre-py/src/main.py (hunk 6 above)
Comment (guumaster) [reply runtime-runners/kre-py/src/main.py:111]: fixed. added missing config

File: runtime-entrypoint/entrypoint.sh (hunk 8 above)
Comment (guumaster) [reply runtime-entrypoint/entrypoint.sh:9]: yes. just uploaded the two greeter versions (go & python) to an empty runtime and both works.

**Pull request merged**